File: android/src/main/java/to/holepunch/bare/android/HostIpcCoordinator.kt
package to.holepunch.bare.android

import android.util.Log
import org.json.JSONException
import org.json.JSONObject
import to.holepunch.bare.kit.IPC
import java.nio.ByteBuffer

/**
 * Handles host-side IPC: capability queries and host invokes. The web layer
 * talks to the worklet over the loopback WebSocket socket, so nothing is
 * relayed to a WebView here.
 */
class HostIpcCoordinator(
    private val ipc: IPC,
    private val hostPlugins: HostPluginRegistry,
) {
    companion object {
        private const val MAX_FIELD_BYTES = 256
    }

    private val decoder = FrameDecoder()

    fun start() {
        ipc.readable {
            try {
                val data = ipc.read() ?: return@readable
                val raw = ByteArray(data.remaining()).also { data.get(it) }
                val messages = decoder.push(raw)
                // A framing violation marks the decoder inert but still returns the
                // frames decoded before it; drop the corrupt byte stream and resync.
                if (decoder.error != null) {
                    if (BuildConfig.DEBUG) Log.e("BARE_KOTLIN", "IPC frame decoder desynced; resetting")
                    decoder.clear()
                }
                for (message in messages) {
                    dispatch(message)
                }
            } catch (e: Exception) {
                // The readable loop must survive any thrown frame so a single
                // malformed/oversized envelope never permanently stalls all host
                // invokes. Logging is debug-only (URLs/tokens may transit here).
                if (BuildConfig.DEBUG) Log.e("BARE_KOTLIN", "Error in readable callback", e)
            }
        }
    }

    private fun dispatch(message: BareProtocol.WireMessage) {
        val headerObj = try {
            JSONObject(message.header)
        } catch (e: JSONException) {
            // Header was not valid JSON: answer with an error frame so the
            // web layer never waits on an unanswered correlation id, then
            // re-arm the readable loop and continue.
            writeError(null, ErrorCodes.FRAME_INVALID, "Invalid envelope header")
            return
        }
        when (headerObj.optString("type")) {
            "HOST_CAPABILITIES_QUERY" -> {
                val reqId = headerObj.optString("requestId")
                if (!validField(reqId)) {
                    writeError(null, ErrorCodes.FRAME_INVALID, "Missing or invalid requestId")
                    return
                }
                val resp = JSONObject().apply {
                    put("type", "HOST_CAPABILITIES_RESPONSE")
                    put("requestId", reqId)
                    put("capabilities", hostPlugins.toCapabilitiesJson())
                }
                ipc.write(
                    BareProtocol.buildMessage(
                        BareProtocol.MessageType.ENVELOPE,
                        resp.toString(),
                        null,
                    ),
                )
            }
            "HOST_INVOKE_REQUEST" -> {
                val reqId = headerObj.optString("requestId")
                val pluginId = headerObj.optString("pluginId")
                val event = headerObj.optString("event")
                if (!validField(reqId) || !validField(pluginId) || !validField(event)) {
                    // Echo the original header (if any) so a correlated error
                    // frame is returned; never let one bad frame stall the loop.
                    writeError(message.header, ErrorCodes.FRAME_INVALID, "Malformed HOST_INVOKE_REQUEST")
                    return
                }
                val args = headerObj.optJSONObject("args")
                hostPlugins.dispatch(pluginId, event, args, message.payload) { outcome ->
                    val respHeader = JSONObject().apply {
                        put("type", "HOST_INVOKE_RESPONSE")
                        put("requestId", reqId)
                        put("pluginId", pluginId)
                        put("event", event)
                    }
                    when (outcome) {
                        is HostPluginRegistry.HostInvokeOutcome.Ok ->
                            respHeader.put("result", outcome.value)
                        is HostPluginRegistry.HostInvokeOutcome.Fail ->
                            respHeader.put(
                                "error",
                                JSONObject().apply {
                                    put("code", outcome.code)
                                    put("message", outcome.message)
                                },
                            )
                    }
                    try {
                        ipc.write(
                            BareProtocol.buildMessage(
                                BareProtocol.MessageType.ENVELOPE,
                                respHeader.toString(),
                                null,
                            ),
                        )
                    } catch (e: Exception) {
                        if (BuildConfig.DEBUG) Log.e("BARE_KOTLIN", "Failed to write invoke response", e)
                    }
                }
            }
            else -> {
                writeError(message.header, ErrorCodes.FRAME_INVALID, "Unknown envelope type")
                return
            }
        }
    }

    private fun writeError(requestHeader: String?, code: String, message: String) {
        try {
            val header = requestHeader ?: "{}"
            val err = BareProtocol.buildErrorResponse(
                header,
                code,
                message,
                "HOST_INVOKE_RESPONSE",
            )
            ipc.write(BareProtocol.buildMessage(err.type, err.header, err.payload))
        } catch (e: Exception) {
            if (BuildConfig.DEBUG) Log.e("BARE_KOTLIN", "Failed to write error frame", e)
        }
    }

    private fun validField(s: String): Boolean = s.length in 1..MAX_FIELD_BYTES
}

File: ios/Sources/BareHost/HostIpcCoordinator.swift
import BareKit
import Foundation

/**
 * Handles host-side IPC: capability queries and host invokes. The web layer
 * talks to the worklet over the loopback WebSocket socket, so nothing is
 * relayed to a WKWebView here. Mirror of
 * `android/.../HostIpcCoordinator.kt`.
 */
public final class HostIpcCoordinator {
  private let ipc: IPC
  private let hostPlugins: HostPluginRegistry

  public init(
    ipc: IPC,
    hostPlugins: HostPluginRegistry
  ) {
    self.ipc = ipc
    self.hostPlugins = hostPlugins
  }

  public func start() {
    Task {
      do {
        for try await chunk in ipc {
          do {
            try await handle(data: chunk)
          } catch {
            // Log, never silently swallow: one bad frame must not kill the loop.
            // No interpolated error value — keep the message static.
            BareHostLogger.log("Error handling IPC frame")
          }
        }
      } catch {
        BareHostLogger.log("IPC stream ended")
      }
    }
  }

  private func handle(data: Data) async throws {
    guard let message = BareProtocol.parseMessage(data) else { return }
    let header = try jsonObject(message.header)
    switch header["type"] as? String ?? "" {
    case "HOST_CAPABILITIES_QUERY":
      let reqId = header["requestId"] as? String ?? ""
      let response: [String: Any] = [
        "type": "HOST_CAPABILITIES_RESPONSE",
        "requestId": reqId,
        "capabilities": hostPlugins.toCapabilitiesJSON(),
      ]
      let frame = try BareProtocol.buildMessage(
        type: BareProtocol.MessageType.envelope,
        headerJson: BareProtocol.encodeJSON(response),
        payload: nil
      )
      try await ipc.write(data: frame)

    case "HOST_INVOKE_REQUEST":
      let reqId = header["requestId"] as? String ?? ""
      let pluginId = header["pluginId"] as? String ?? ""
      let event = header["event"] as? String ?? ""
      let args = header["args"] as? [String: Any]

      // `respond` must fire exactly once: the registry handler may respond
      // asynchronously (e.g. after a native picker), and a handler that never
      // calls back would otherwise strand the web-side promise forever.
      let responder = SingleResponse { [weak self] outcome in
        var responseHeader: [String: Any] = [
          "type": "HOST_INVOKE_RESPONSE",
          "requestId": reqId,
          "pluginId": pluginId,
          "event": event,
        ]
        switch outcome {
        case .ok(let value):
          responseHeader["result"] = value
        case .fail(let code, let message):
          responseHeader["error"] = [
            "code": code,
            "message": message,
          ]
        }
        let frame = try? BareProtocol.buildMessage(
          type: BareProtocol.MessageType.envelope,
          headerJson: BareProtocol.encodeJSON(responseHeader),
          payload: nil
        )
        if let frame {
          Task { [weak self] in
            try? await self?.ipc.write(data: frame)
          }
        }
      }
      // Invoke deadline: emit a canonical timeout if the handler never responds.
      Task {
        try? await Task.sleep(nanoseconds: 30_000_000_000)
        responder.call(.fail(code: ErrorCodes.timeout, message: "Host invoke timed out"))
      }
      hostPlugins.dispatch(
        pluginId: pluginId,
        event: event,
        args: args,
        payload: message.payload,
        respond: responder.call
      )

    default:
      // Unknown header type: never silently drop the frame — answer with a
      // canonical HOST_ERROR so the web side can surface a typed failure.
      let reqId = header["requestId"] as? String ?? ""
      var responseHeader: [String: Any] = [
        "type": "HOST_INVOKE_RESPONSE",
        "requestId": reqId,
      ]
      responseHeader["error"] = [
        "code": ErrorCodes.hostError,
        "message": "Unknown IPC header type",
      ]
      if let frame = try? BareProtocol.buildMessage(
        type: BareProtocol.MessageType.envelope,
        headerJson: BareProtocol.encodeJSON(responseHeader),
        payload: nil
      ) {
        try? await ipc.write(data: frame)
      }
    }
  }

  /// Wraps a response callback so it runs at most once (success, failure, or
  /// invoke timeout all race to call it; only the first wins).
  private final class SingleResponse {
    private let lock = NSLock()
    private var done = false
    private let send: (HostPluginRegistry.HostInvokeOutcome) -> Void

    init(_ send: @escaping (HostPluginRegistry.HostInvokeOutcome) -> Void) {
      self.send = send
    }

    func call(_ outcome: HostPluginRegistry.HostInvokeOutcome) {
      lock.lock()
      defer { lock.unlock() }
      guard !done else { return }
      done = true
      send(outcome)
    }
  }

  private func jsonObject(_ header: String) throws -> [String: Any] {
    let object =
      try JSONSerialization.jsonObject(with: Data(header.utf8))
    guard let dictionary = object as? [String: Any] else {
      throw HostIpcError.invalidHeader
    }
    return dictionary
  }

  private enum HostIpcError: Error {
    case invalidHeader
  }
}
